### PR TITLE
DTPools 2.0

### PR DESCRIPTION
**Pull request by pavanbalaji**
New version of the DTPools library.

Most of the DTPools 1.0 functionality has been ported over to the new code base, and a lot of new cases are also now covered.  We are planning to add some more functionality before this PR is ready.

Only one test has been ported over to the new interface, so the remaining tests will fail to build.

This PR depends on #3720, which is a bug in MPICH and is triggered by this new framework.

### CODE REVIEW

File: src/mpi/datatype/type_indexed.c [hunk 1]
```diff
@@ -143,6 +143,8 @@ int MPIR_Type_indexed(int count,
 
         new_dtp->has_sticky_lb = old_dtp->has_sticky_lb;
         new_dtp->has_sticky_ub = old_dtp->has_sticky_ub;
+
+        new_dtp->alignsize = old_dtp->alignsize;
```
Comment (gcongiu) [inline src/mpi/datatype/type_indexed.c]: shouldn't be this a separate PR

Comment (pavanbalaji) [reply src/mpi/datatype/type_indexed.c]: It is.  That got merged in.  I need to rebase on to `mpich/master`.

File: test/mpi/dtpools/src/dtpools_misc.c [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    DTPI_FUNC_EXIT();
+    return count / DTPI_low_count(count);
+}
+
+void DTPI_reset_rand(DTPI_pool_s * dtpi)
+{
+    DTPI_FUNC_ENTER();
+
+    srand(dtpi->seed);
+    for (int i = 0; i < dtpi->rand_count; i++)
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/src/dtpools_misc.c]: Done.

File: test/mpi/dtpools/src/dtpools_misc.c [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    int i;
+    MPI_Aint lb;
+    int rc = DTP_SUCCESS;
+
+    DTPI_FUNC_ENTER();
+
+    DTPI_ERR_ASSERT(str, rc);
+
+    DTPI_ALLOC_OR_FAIL(dtpi->base_type_str, MAX_TYPE_LEN, rc);
+    strncpy(dtpi->base_type_str, str, strlen(str) + 1);
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/src/dtpools_misc.c]: done.

File: test/mpi/dtpools/README [hunk 4]
```diff
@@ -76,60 +68,72 @@ Follows a simple test application that uses DTPools:
...
                 /* error handling */;
             }
 
-            MPI_Irecv(recv_dtp->DTP_obj_array[j].DTP_obj_buf,
-                      recv_dtp->DTP_obj_array[j].DTP_obj_count,
-                      recv_dtp->DTP_obj_array[j].DTP_obj_type,
+            MPI_Irecv(recv_obj.DTP_obj_buf,
+                      recv_obj.DTP_obj_count,
+                      recv_obj.DTP_obj_type,
                       myrank, 0, MPI_COMM_WORLD, &req);
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/README]: Removed the example in the README.

File: test/mpi/dtpools/README [hunk 5]
```diff
@@ -76,60 +68,72 @@ Follows a simple test application that uses DTPools:
...
 
-            MPI_Send(send_dtp->DTP_obj_array[i].DTP_obj_buf,
-                     send_dtp->DTP_obj_array[i].DTP_obj_count,
-                     send_dtp->DTP_obj_array[i].DTP_obj_type,
+            MPI_Send(send_obj.DTP_obj_buf,
+                     send_obj.DTP_obj_count,
+                     send_obj.DTP_obj_type,
                      myrank, 0, MPI_COMM_WORLD);
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/README]: Removed the example in the README.

File: test/mpi/dtpools/README [hunk 6]
```diff
@@ -76,60 +68,72 @@ Follows a simple test application that uses DTPools:
...
 
             MPI_Wait(&req, MPI_STATUS_IGNORE);
 
-            if (DTP_obj_buf_check(recv_dtp, j, 0, 2, basic_type_count) != DTP_SUCCESS) {
+            if (DTP_obj_buf_check(recv_obj, 0, 2, base_type_count) != DTP_SUCCESS) {
                 num_err++;
             }
-            DTP_obj_free(recv_dtp, j);
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/README]: Removed the example in the README.

File: test/mpi/dtpools/README [hunk 7]
```diff
@@ -76,60 +68,72 @@ Follows a simple test application that uses DTPools:
...
 
-        for (j = 0; j < recv_dtp->DTP_num_objs; j++) {
-            err = DTP_obj_create(recv_dtp, j, 0, 0, 0);
-            if (err != DTP_SUCCESS) {
+        for (j = 0; j < NUM_OBJECTS_TO_TEST; j++) {
+            rc = DTP_obj_create(recv_dtp, &recv_obj);
+            if (rc != DTP_SUCCESS) {
+                /* error handling */;
+            }
+
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/README]: Removed the example in the README.

File: test/mpi/dtpools/src/dtpools_init_verify.c [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+
+  fn_fail:
+    goto fn_exit;
+}
+
+int DTPI_init_verify(DTP_pool_s dtp, DTP_obj_s obj, DTPI_Attr_s * attr_tree,
+                     size_t buf_offset, int *val_start, int val_stride, int *rem_val_count,
+                     int verify)
+{
+    int rc = DTP_SUCCESS;;
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/src/dtpools_init_verify.c]: fixed.

File: test/mpi/dtpools/configure.ac [hunk 9]
```diff
@@ -50,21 +53,6 @@ AC_ARG_VAR([DTP_MIN_OBJ_ID],
 AC_ARG_VAR([DTP_MAX_OBJ_ID],
```
Comment (gcongiu) [inline test/mpi/dtpools/configure.ac]: I forgot about `DTP_MIN/MAX_OBJ_ID`. We are no longer using these. We should probably remove them from the configure

**Comment by pavanbalaji** [comment]
This PR is currently failing because of #3749.

File: test/mpi/dtpools/configure.ac (hunk 9 above)
Comment (pavanbalaji) [reply test/mpi/dtpools/configure.ac]: Fixed.

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by yfguo** [comment]
I think all the prerequisites for this PR that Guiseppe has mentioned are merged.

**Comment by gcongiu** [comment]
@yfguo https://github.com/pmodels/mpich/pull/3756 still not merged

**Comment by raffenet** [comment]
@gcongiu can older dtpools enhancement PRs be closed? E.g. #3709, #3678, ...

**Comment by gcongiu** [comment]
@raffenet I think so. We can still reopen them in the unlikely case we will need them again.

**Comment by pavanbalaji** [comment]
There’s at least one more bug in ch3/tcp, that Min and I are working on.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by yfguo** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by yfguo** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by yfguo** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
@yfguo why did you restart the tests?

**Comment by yfguo** [comment]
I was just trying to get a clean set of results after changing the job time.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
Tests look fine.  Retesting one more time after rebase.

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
Tests look good, except for the `gather_big` failures in `ucx` that are being handled in #3770.

This is ready for your review, @gcongiu and @yfguo.

**Comment by pavanbalaji** [comment]
I had forgotten that Yanfei is on vacation this week.  Pinging @wesbland to see if he can review this.

File: test/mpi/configure.ac [hunk 10]
```diff
@@ -251,6 +253,47 @@ AC_ARG_ENABLE(collalgo-tests,
                               [Run extended collective algorithm tests])],,
               [enable_collalgo_tests=no])
 
+# DTPools test generation
+AC_ARG_WITH(dtpools-datatypes,
+    [AC_HELP_STRING([--with-dtpools-datatypes=typelist],
+        [Semicolon separated list of MPI datatypes to use for DTPools tests
```
Comment (gcongiu) [inline test/mpi/configure.ac]: This is actually my fault. Datatypes are coma separated not semicolon separated.

File: test/mpi/gentests.sh [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+srcdir=`dirname $0`
+builddir=$PWD
+
+# reformat type list
+types=$(echo $types | tr "," " ")
+
+# set max count and large count
+MAXSIZE=1000
+MAXPINGSIZE=100
+MAXCOUNT=262144
```
Comment (gcongiu) [inline test/mpi/gentests.sh]: These three variables are no longer needed. I can remove them if you want.

File: test/mpi/gentests.sh [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        mintestsize=`echo $line | cut -f6 -d':'`
+        maxtestsize=`echo $line | cut -f7 -d':'`
+        other_args=""
+    else
+        # the line is a comment
+        continue
+    fi
+
+    if [ -z "$pathname" ] ; then
+	echo "No pathname found"
```
Comment (gcongiu) [inline test/mpi/gentests.sh]: It looks like in my editor indentation is all messed up. Can you use spaces instead of tabs? Again, if you need I can fix this.

File: test/mpi/configure.ac [hunk 13]
```diff
@@ -251,6 +253,47 @@ AC_ARG_ENABLE(collalgo-tests,
...
+AC_ARG_WITH(dtpools-maxbufsize,
+    [AC_HELP_STRING([--with-dtpools-maxbufsize=size_in_bytes],
+        [maximum buffer size needed for the generated datatypes.])],,)
+
+AC_CHECK_SIZEOF(void *)
+if test ! -z "${with_dtpools_maxbufsize}" ; then
+    # if the user specified a size, use it; otherwise try to
+    # autodetect it
+    dtp_args="$dtp_args --with-dtpools-maxbufsize=${with_dtpools_maxbufsize}"
+    :
```
Comment (gcongiu) [inline test/mpi/configure.ac]: extra `:`?

**Comment by pavanbalaji** [comment]
Found a few small issues while reviewing the PR myself.  These fixes could have easily gone in as a separate PR, but since no one has started reviewing yet, I figured I'll add them to this PR too.

test:jenkins/ch3/most
test:jenkins/ch4/most

File: test/mpi/dtpools/src/dtpools_attr.c [hunk 14]
```diff
@@ -0,0 +1,1488 @@
...
+                max_displ_idx = i;
+            else if (attr->u.vector.stride > 0)
+                max_displ_idx = i;
+
+            if (min_displ_idx == -1)
+                min_displ_idx = i;
+            else if (attr->u.vector.stride < 0)
+                min_displ_idx = i;
+        }
+
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/src/dtpools_attr.c]: Those are the indices for the minimum and maximum displacement.  Those need not be `0` and `numblks-1`.  For example, an intermediate displacement could have the smallest value.

File: test/mpi/dtpools/src/dtpools_attr.c [hunk 15]
```diff
@@ -0,0 +1,1488 @@
...
+    MPI_Aint tmp_lb;
+    rc = MPI_Type_get_extent(type, &tmp_lb, &attr->child_type_extent);
+    DTPI_ERR_CHK_MPI_RC(rc);
+
+    MPI_Aint tmp_count = count;
+    attr->u.indexed.array_of_blklens = NULL;
+    attr->u.indexed.array_of_displs = NULL;
+    while (1) {
+      retry:
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/src/dtpools_attr.c]: Yeah, I guess I can do that.

File: test/mpi/configure.ac (hunk 10 above)
Comment (pavanbalaji) [reply test/mpi/configure.ac]: I'll fix it.

File: test/mpi/gentests.sh (hunk 11 above)
Comment (pavanbalaji) [reply test/mpi/gentests.sh]: These have been removed in a later fixup commit.

File: test/mpi/gentests.sh (hunk 12 above)
Comment (pavanbalaji) [reply test/mpi/gentests.sh]: I'll fix it.

File: test/mpi/configure.ac (hunk 13 above)
Comment (pavanbalaji) [reply test/mpi/configure.ac]: Yup, I'll fix it.

File: test/mpi/dtpools/src/dtpools_attr.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply test/mpi/dtpools/src/dtpools_attr.c]: Oh, I remember why I did it this way.  The "goto retry" was happening inside nested loop.

**Comment by pavanbalaji** [comment]
Addressed @gcongiu's comments.

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by gcongiu** [comment]
@wesbland I am happy with the current state of this PR. All my previous comments have been addressed/clarified and I don't think I have any further comment. If you want you can start your review.

**Comment by pavanbalaji** [comment]
@gcongiu I had re-requested a review from you because I thought you were planning to review again.  If that's not the case, can you go ahead and mark the PR as approved again?

**Comment by gcongiu** [comment]
@pavanbalaji I was missing some updates in the test suite. I looked at them today. Every thin seems fine.
